compiler/mir/mirgen.nim

```diff
@@ -0,0 +1,2105 @@
+  ## Tests and returns whether the simple or compound statement `n` ends in a
+  ## no-return statement
+
+  # TODO: this is a patched version of ``sem.endsInNoReturn`` that also
+  #       considers ``nkPragmaBlock``. Move this procedure somewhere common and
+  #       replace ``sem.endsInNoReturn`` with it
+  const SkipSet = {nkStmtList, nkStmtListExpr, nkPragmaBlock}
+  var it {.cursor.} = n
+  while it.kind in SkipSet and it.len > 0:
+    it = it.lastSon
+
+  result = it.kind in nkLastBlockStmts or
+    (it.kind in nkCallKinds and it[0].kind == nkSym and
+     sfNoReturn in it[0].sym.flags)
+
+func canonicalExpr(n: PNode): PNode =
+  ## Returns the canonical expression, i.e. the expression without leading
+  ## pragma blocks or empty statement lists
+  func skipped(n: PNode): PNode =
+    case n.kind
+    of nkPragmaBlock:
+      result = n.lastSon
+    of nkStmtListExpr:
+      # see if all nodes except the last one are empty. If they are, the
+      # ``nkStmtListExpr`` is skipped
+      for i in 0..<n.len-1:
+        if n[i].kind != nkEmpty:
+          # there are statements in the list -> the list needs to be kept
+          return n
+
+      result = n.lastSon
+    else:
+      result = n
+
+  var n {.cursor.} = n
+  while (let it = skipped(n); it != n):
+    n = it
+
+  result = n
+
+func selectWhenBranch(n: PNode, isNimvm: bool): PNode =
+  assert n.kind == nkWhen
+  if isNimvm: n[0][1]
+  else:       n[1][0]
+
+func selectDefKind(s: PSym): range[mnkDef..mnkDefCursor] =
+  ## Returns the kind of definition to use for the given entity
+  if sfCursor in s.flags: mnkDefCursor
+  else:                   mnkDef
+
+func isSome(x: Destination): bool {.inline.} =
+  x.kind != dkNone
+
+func initDestination(n: PNode, isFirst: bool): Destination =
+  result = Destination(kind: dkGen, node: n, isEmpty: isFirst, own: not isCursor(n))
+
+proc typeOrVoid(g: ModuleGraph, t: PType): PType =
+  ## Returns `t` if it's not 'nil' - the ``void`` type otherwise
+  if t != nil: t
+  else:        g.getSysType(unknownLineInfo, tyVoid)
+
+proc typeOrVoid(c: TCtx, t: PType): PType {.inline.} =
+  ## Returns `t` if it's not 'nil' - the ``void`` type otherwise
+  # TODO: cache the void type
+  typeOrVoid(c.graph, t)
+
+func tempNode(typ: PType, id: TempId): MirNode {.inline.} =
+  MirNode(kind: mnkTemp, typ: typ, temp: id)
+
+
+func nextTempId(c: var TCtx): TempId =
+  result = TempId(c.numTemps)
+  inc c.numTemps
+
+func nextLabel(c: var TCtx): LabelId =
+  result = LabelId(c.numLabels + 1)
+  inc c.numLabels
+
+# ----------- CodeFragment API -------------
+
+template add(f: var CodeFragment, n: MirNode) =
+  f.nodes.add n
+
+template subTree(f: var CodeFragment, n: MirNode, body: untyped) =
+  f.nodes.subTree n:
+    body
+
+func apply(frag: var CodeFragment, id: SourceIndex) =
+  ## Associates all nodes added since the previous call to ``apply`` with the
+  ## origin information identified by `id`
+  assert frag.nodes.len > frag.source.len
```

Comment:
You should drop this assert. For something like this I consider doing nothing a perfectly valid scenario.

Reply:
Calling `apply` when there's nothing to update hints at a logic or programming error somewhere, hence the assert. I would prefer to keep it (but can update the documentation).